03_oop_part_2/src/item.ts

```diff
@@ -2,32 +2,46 @@ import { Page } from './page';
 import { Pages } from './pages';
 
 
-const PagesIterable = () => class implements IterableIterator<Item> {
-    constructor(protected pages: Pages, protected openPage?: Page) {}
-
-    public next(): IteratorResult<Item> {
-        let nxt = this.pages.next();
-        if (!nxt.done) {
-            this.openPage = nxt.value;
-            return {done: false, value: this};
+export abstract class Item {
+
+    public abstract toString(): String;
+}
+
+type Constructor = new (...args: any[]) => {};
+
+export function PagesIterable<ItemType extends Constructor>(TheItem: ItemType) {
```

Comment:
Another approach is to move `protected pages` to `Item` class and add constraints to the `Constructor` so that it instantiates `Item` class instances. Thus we can access `pages` property in derived classes (`Book`, etc) and we don't need `setPages` method in `Mixed` class

Reply:
Great, thanks!